rust/feature-flags/src/flag_matching.rs

```diff
@@ -132,90 +288,144 @@ impl FeatureFlagMatcher {
         _index: usize,
     ) -> Result<(bool, String), FlagError> {
         let rollout_percentage = condition.rollout_percentage.unwrap_or(100.0);
-        if let Some(properties) = &condition.properties {
-            if properties.is_empty() {
-                return Ok(self.check_rollout(feature_flag, rollout_percentage));
+        if let Some(flag_property_filters) = &condition.properties {
+            if flag_property_filters.is_empty() {
+                return self.check_rollout(feature_flag, rollout_percentage).await;
             }
 
-            let target_properties = self.get_target_properties(feature_flag, properties).await?;
+            let target_properties =
+                if let Some(group_type_index) = feature_flag.get_group_type_index() {
+                    self.get_or_fetch_group_properties(
+                        feature_flag,
+                        group_type_index,
+                        flag_property_filters,
+                    )
+                    .await?
+                } else {
+                    self.get_or_fetch_person_properties(feature_flag.team_id, flag_property_filters)
+                        .await?
+                };
 
-            if !self.all_properties_match(properties, &target_properties) {
+            if !self.all_properties_match(flag_property_filters, &target_properties) {
                 return Ok((false, "NO_CONDITION_MATCH".to_string()));
             }
         }
 
-        Ok(self.check_rollout(feature_flag, rollout_percentage))
+        self.check_rollout(feature_flag, rollout_percentage).await
     }
 
-    async fn get_target_properties(
+    async fn get_or_fetch_group_properties(
```

Comment:
so this is different than django because the way django worked is that it fetched properties without ever touching the DB, but then handled empty properties by generating them as part of building the queries that it used for condition matching.  

Given that our new approach is to explicitly differentiate between fetching properties from the DB and then matching them, this method now talks to Postgres if it can't get flags from the payload or the cache.

Reply:
eventually, I think we should minimise this to max two DB calls as well: Analyse all flags, and fetch person properties + ALL group properties if need be.

Right now it goes to the db for each group type index. Not sure if the schemas make this more annoying to do, (I don't think so?) - persons & groups together would be more annoying and probably slower though.

Reply:
> eventually, I think we should minimise this to max two DB calls as well: Analyse all flags, and fetch person properties + ALL group properties if need be.

Oh you're so right, I was thinking of this too much from the perspective of just one flag.  So the idea here is that, once we've figured out what properties we need to evaluate from _all_ the flags, then get the data needed.  I can do that!